File: .orkestrel/campaign/docs-parity/instruments/d7/units/websocket/d7n-websocket-converge-fix/shape.py
from pathlib import Path
import re

p = Path('guides/websocket.md')
lines = p.read_text(encoding='utf8').split('\n')

CONSTANT_SHAPES = {
    'WEBSOCKET_GUID': '`string`',
    'WEBSOCKET_VERSION': '`string`',
    'WEBSOCKET_OPCODE_TEXT': '`number`',
    'WEBSOCKET_OPCODE_BINARY': '`number`',
    'WEBSOCKET_OPCODE_CONTINUATION': '`number`',
    'WEBSOCKET_OPCODE_CLOSE': '`number`',
    'WEBSOCKET_OPCODE_PING': '`number`',
    'WEBSOCKET_OPCODE_PONG': '`number`',
    'WEBSOCKET_READY_CONNECTING': '`WebSocketReadyState`',
    'WEBSOCKET_READY_OPEN': '`WebSocketReadyState`',
    'WEBSOCKET_READY_CLOSING': '`WebSocketReadyState`',
    'WEBSOCKET_READY_CLOSED': '`WebSocketReadyState`',
    'WEBSOCKET_CLOSE_NORMAL': '`number`',
    'WEBSOCKET_CLOSE_PROTOCOL': '`number`',
    'WEBSOCKET_CLOSE_UNSUPPORTED': '`number`',
    'WEBSOCKET_CLOSE_INVALID': '`number`',
    'WEBSOCKET_CLOSE_TOO_BIG': '`number`',
    'WEBSOCKET_MAX_PAYLOAD': '`number`',
    'WEBSOCKET_CLOSE_TIMEOUT_MS': '`number`',
    'WEBSOCKET_CONTROL_MAX_LENGTH': '`number`',
    'WEBSOCKET_CLOSE_REASON_MAX_LENGTH': '`number`',
    'WEBSOCKET_FAIL_TIMEOUT_MS': '`number`',
}

TYPE_SHAPES = {
    'WebSocketReadyState': r'`0 \| 1 \| 2 \| 3`',
    'WebSocketFrame': '`{ fin, opcode, payload, consumed, masked, rsv }`',
    'WebSocketEncodeOptions': '`{ masked?, mask? }`',
    'WebSocketErrorCode': r"`'OPTION' \| 'LIMIT' \| 'CLOSE' \| 'FRAME'`",
    'NodeWebSocketEventMap': '`{ open, message, close, error, ping, pong }`',
    'NodeWebSocketOptions': '`{ socket, key?, head?, protocol?, on?, error?, payload?, timeout?, signal? }`',
    'NodeWebSocketInterface': '`{ emitter, readyState } plus send, ping, close, destroy`',
}

CONSTANTS_SENTENCE = 'A `Shape` cell holds the constant\'s declared type.'
TYPES_SENTENCE = (
    "A `Shape` cell holds an interface's data members as bare names in braces, `?` marking an "
    "optional member and `plus` introducing its call-signature members, and a type alias's own "
    "type literal with a union's arms escaped as `\\|`."
)


def split_row(line):
    assert line.startswith('|') and line.endswith('|'), line
    return [cell.strip() for cell in line[1:-1].split('|')]


def render(rows):
    widths = [max(len(row[index]) for row in rows) for index in range(len(rows[0]))]
    out = []
    for position, row in enumerate(rows):
        cells = [row[index].ljust(widths[index]) for index in range(len(row))]
        out.append('| ' + ' | '.join(cells) + ' |')
        if position == 0:
            out.append('| ' + ' | '.join('-' * widths[index] for index in range(len(row))) + ' |')
    return out


def rebuild(heading, shapes, sentence):
    start = lines.index(heading)
    first = next(index for index in range(start, len(lines)) if lines[index].startswith('|'))
    last = first
    while last + 1 < len(lines) and lines[last + 1].startswith('|'):
        last += 1
    table = [split_row(line) for line in lines[first:last + 1]]
    header, body = table[0], table[2:]
    assert header[-1] == 'Summary', header
    rows = [header[:2] + ['Shape'] + header[2:]]
    for row in body:
        name = row[0].strip('`')
        assert name in shapes, name
        rows.append(row[:2] + [shapes[name]] + row[2:])
    lines[first:last + 1] = render(rows)
    lines.insert(first, '')
    lines.insert(first, sentence)


rebuild('### Types', TYPE_SHAPES, TYPES_SENTENCE)
rebuild('### Constants', CONSTANT_SHAPES, CONSTANTS_SENTENCE)

p.write_text('\n'.join(lines), encoding='utf8')
print('shape ok')

File: .orkestrel/campaign/docs-parity/instruments/d7/units/sea/d7n-sea-converge/shape.mjs
// Inserts the `Shape` column into the named tables of guides/sea.md, between
// `Kind` and `Summary`, and writes the convention sentence above each table.
// Splits a row on a pipe not preceded by a backslash, so a `\|` inside a union
// literal stays inside its cell. Every non-`Summary` cell keeps its text.
import { readFileSync, writeFileSync } from 'node:fs'

const PATH = 'guides/sea.md'
const SPLIT = /(?<!\\)\|/

const CONSTANT_SENTENCE = 'A `Shape` cell holds the constant’s declared type.'.replace(
	'’',
	"'",
)
const TYPE_SENTENCE =
	"A `Shape` cell holds an interface's data members as bare names in braces, `?` marking an optional member and `plus` introducing its call-signature members, and a type alias's own type literal with a union's arms escaped as `\\|`."

const CONSTANTS = {
	SEA_SENTINEL_FUSE: '`string`',
	SEA_BLOB_RESOURCE: '`string`',
	DEFAULT_SEA_COMPRESSION_QUALITY: '`number`',
	WINDOWS_SUBSYSTEM_GUI: '`number`',
	WINDOWS_SUBSYSTEM_CONSOLE: '`number`',
	BROTLI_EXTENSION: '`string`',
	SKIP_EXTENSIONS: '`ReadonlySet<string>`',
	PE_MAGIC: '`number`',
	PE_SIGNATURE: '`number`',
	PE32_MAGIC: '`number`',
	PE32_PLUS_MAGIC: '`number`',
	ELF_MAGIC: '`number`',
	ELF_CLASS_64: '`number`',
	ELF_DATA_LSB: '`number`',
	ELF_PT_NOTE: '`number`',
	ELF_PT_LOAD: '`number`',
	ELF_PT_PHDR: '`number`',
	ELF_PF_R: '`number`',
	ELF_PAGE_SIZE: '`number`',
	MACHO_MAGIC_64: '`number`',
	MACHO_LC_SEGMENT_64: '`number`',
	PE_RT_RCDATA: '`number`',
	PE_RESOURCE_DIR_SIZE: '`number`',
	PE_RESOURCE_ENTRY_SIZE: '`number`',
	PE_RESOURCE_DATA_ENTRY_SIZE: '`number`',
	PE_SECTION_HEADER_SIZE: '`number`',
	PE_RESOURCE_SUBDIR_FLAG: '`number`',
	PE_RESOURCE_NAME_FLAG: '`number`',
	PE_SCN_INITIALIZED_DATA: '`number`',
	PE_SCN_MEM_READ: '`number`',
	SEA_PLATFORMS: '`Readonly<Record<string, SEAPlatform>>`',
	SEA_COMPRESSION_MODE_VALUES: '`Readonly<Record<SEACompressionMode, number>>`',
	DEFAULT_ENTRY_FORMAT: '`SEAEntryFormat`',
}

const TYPES = {
	SEACompressionSize: '`{ original, compressed, ratio }`',
	SEACompressionMode: "`'generic' \\| 'text' \\| 'font'`",
	SEACompressionResult: '`{ input, output, size }`',
	SEACompressionManifest: '`{ assets, total }`',
	SEAProgress: '`{ path, current, total }`',
	SEACompressionHandler: '`(result: SEACompressionResult) => void`',
	SEABrotliOptions: '`{ mode?, quality? }`',
	SEACompressionOptions: '`{ paths, mode?, quality? }`',
	SEAPlatform: '`{ executable, remove?, sign?, verify? }`',
	SEAShellOptions: '`{ cwd?, env?, timeout?, signal? }`',
	ExecutableFormat: "`'pe' \\| 'elf' \\| 'macho'`",
	ELFNoteHeader: '`{ header, total }`',
	ELFProgramHeader: '`{ type, flags, offset, vaddr, paddr, filesz, memsz, align }`',
	PEResourceLeaf:
		'`{ typeId, typeName, nameId, nameName, language, codePage, dataRVA, dataSize }`',
	PEResourceEntry: '`{ language, codePage, leafIndex, dataSize }`',
	PESection:
		'`{ name, virtualSize, virtualAddress, rawSize, rawOffset, characteristics, headerOffset }`',
	InjectorOptions: '`{ executable, resource, blob, fuse?, macho?, overwrite? }`',
	InjectorMachOOptions: '`{ segment? }`',
	InjectorInterface: '`{ format } plus inject`',
	AssetInput: '`{ key, content, compressed? }`',
	AssetInterface: '`{ key, content, compressed }`',
	AssetManagerEventMap: '`{ register, load, clear, error }`',
	AssetManagerOptions: '`{ on?, error?, root?, assets? }`',
	AssetManagerInterface:
		'`{ emitter, count } plus asset, assets, keys, register, load, clear, destroy`',
	SEAStatus: "`'idle' \\| 'active' \\| 'done' \\| 'error'`",
	SEAErrorCode:
		"`'PLATFORM' \\| 'ENTRY' \\| 'ASSET' \\| 'BLOB' \\| 'FORMAT' \\| 'INJECT' \\| 'ROOM' \\| 'FUSE' \\| 'SIGN' \\| 'SHELL' \\| 'TIMEOUT' \\| 'ABORT' \\| 'OUTPUT' \\| 'STATE' \\| 'BROWSER'`",
	SEAEntryFormat: "`'cjs' \\| 'esm'`",
	SEAEntryOptions: '`{ path, format? }`',
	SEABlobOptions: '`{ cache?, snapshot? }`',
	SEAEventMap: '`{ compress, progress, blob, assemble, complete, error }`',
	SEAOptions:
		'`{ on?, error?, name, entry, output, assets?, compression?, windows?, root?, signal?, timeout?, blob? }`',
	SEAWindowsOptions: '`{ terminal?, sign? }`',
	SEAWindowsSignOptions: '`{ file?, password?, thumbprint?, timestamp?, digest? }`',
	SEAResult:
		'`{ executable, platform, size, duration, compression?, signed, stripped, terminal? }`',
	SEAInterface: '`{ emitter, status } plus execute, destroy`',
}

/**
 * Splits one table row into its cells.
 *
 * @param line - The row, with its leading and trailing pipe.
 * @returns One entry per cell, trimmed.
 */
function splitRow(line) {
	const cells = line.split(SPLIT)
	return cells.slice(1, cells.length - 1).map((cell) => cell.trim())
}

/**
 * Rebuilds one table row from its cells.
 *
 * @param cells - The cells, in order.
 * @returns The row, with its leading and trailing pipe.
 */
function joinRow(cells) {
	return `| ${cells.join(' | ')} |`
}

/**
 * Inserts one column into every row of the table starting at `start`.
 *
 * @param lines - Every line of the guide.
 * @param start - The index of the table's header row.
 * @param header - The inserted column's header text.
 * @param values - Each row key mapped to its inserted cell.
 * @returns The index one past the table's last row.
 */
function insertColumn(lines, start, header, values) {
	let at = start
	while (at < lines.length && lines[at].startsWith('|')) {
		const cells = splitRow(lines[at])
		if (at === start) {
			cells.splice(2, 0, header)
		} else if (at === start + 1) {
			cells.splice(2, 0, '---')
		} else {
			const key = cells[0].replaceAll('`', '')
			const value = values[key]
			if (value === undefined) throw new Error(`no Shape for ${key}`)
			cells.splice(2, 0, value)
		}
		lines[at] = joinRow(cells)
		at += 1
	}
	return at
}

const lines = readFileSync(PATH, 'utf8').split('\n')
for (const [heading, header, values, sentence] of [
	['### Constants', 'Shape', CONSTANTS, CONSTANT_SENTENCE],
	['### Types', 'Shape', TYPES, TYPE_SENTENCE],
]) {
	const at = lines.indexOf(heading)
	if (at === -1) throw new Error(`no heading ${heading}`)
	let start = at + 1
	while (!lines[start].startsWith('|')) start += 1
	insertColumn(lines, start, header, values)
	lines.splice(start, 0, sentence, '')
}
writeFileSync(PATH, lines.join('\n'))
process.stdout.write('rewrote the Constants and Types tables\n')
